File: csharp/1171.remove-zero-sum-consecutive-nodes-from-linked-list.cs
/*
 * @lc app=leetcode id=1171 lang=csharp
 *
 * [1171] Remove Zero Sum Consecutive Nodes from Linked List
 */

using System.Collections.Generic;

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode RemoveZeroSumSublists(ListNode head) {
        var sentinel = new ListNode();
        var map = new Dictionary<int, ListNode>();
        int acc = 0;
        sentinel.next = head;
        map[0] = sentinel;
        while (head != null) {
            acc += head.val;
            if (map.TryGetValue(acc, out var p)) {
                var rm = p.next;
                int b = acc + rm.val;
                while (b != acc) {
                    map.Remove(b);
                    rm = rm.next;
                    b += rm.val;
                }
                p.next = head.next;
            } else {
                map[acc] = head;
            }
            head = head.next;
        }
        return sentinel.next;
    }
}
// @lc code=end

File: java/1171.remove-zero-sum-consecutive-nodes-from-linked-list.java
/*
 * @lc app=leetcode id=1171 lang=java
 *
 * [1171] Remove Zero Sum Consecutive Nodes from Linked List
 */

import java.util.HashMap;
import java.util.Map;

// @lc code=start

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        Map<Integer, ListNode> sumToNode = new HashMap<>();
        ListNode sentinel = new ListNode(0);
        sumToNode.put(0, sentinel);
        sentinel.next = head;
        ListNode curr = head;
        int sum = 0;
        while (curr != null) {
            sum += curr.val;
            ListNode existing = sumToNode.putIfAbsent(sum, curr);
            if (existing != null) {
                ListNode toRemove = existing.next;
                int sumToRemove = sum;
                while (toRemove != curr) {
                    sumToRemove += toRemove.val;
                    sumToNode.remove(sumToRemove);
                    toRemove = toRemove.next;
                }
                existing.next = curr.next;
            }
            curr = curr.next;
        }
        return sentinel.next;
    }
}
// @lc code=end

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
